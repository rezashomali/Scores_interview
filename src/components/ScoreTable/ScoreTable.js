import React, { useState, useMemo } from "react";
import PropTypes from "prop-types";
import { makeStyles } from "@material-ui/core/styles";
import {
  Link,
  Table,
  Button,
  TableBody,
  TableCell,
  TableHead,
  TableRow,
  Grid,
  TextField,
} from "@material-ui/core";
import ArrowDownwardIcon from "@material-ui/icons/ArrowDownward";
import ArrowUpwardIcon from "@material-ui/icons/ArrowUpward";
import Title from "../Title/Title";

const useStyles = makeStyles((theme) => ({
  seeMore: {
    marginTop: theme.spacing(3),
  },
}));

const ScoreTable = ({ userData }) => {
  const classes = useStyles();
  const [readmore, setReadmore] = useState(false);
  const [searchValue, setSearchValue] = useState("");
  const [sortConfig, setSortConfig] = useState({
    key: null,
    direction: null,
  });

  let sortedData = userData;
  useMemo(() => {
    if (sortConfig.key !== null) {
      sortedData.sort((a, b) => {
        if (
          a[sortConfig.key]?.toLowerCase() < b[sortConfig.key]?.toLowerCase()
        ) {
          return sortConfig.direction === "ascending" ? -1 : 1;
        }
        if (
          a[sortConfig.key]?.toLowerCase() > b[sortConfig.key]?.toLowerCase()
        ) {
          return sortConfig.direction === "ascending" ? 1 : -1;
        }
        return 0;
      });
    }

    return sortedData;
  }, [sortConfig]);

  const requestSort = (key) => {
    let direction = "ascending";
    if (sortConfig.key === key && sortConfig.direction === "ascending") {
      direction = "descending";
    }
    setSortConfig({ key, direction });
  };

  return (
    <React.Fragment>
      <Grid container spacing={3}>
        <Grid item xs={6}>
          <Title>Scores listing</Title>
        </Grid>
        <Grid item xs={6} align="right">
          <TextField
            label="search..."
            variant="outlined"
            size="small"
            value={searchValue}
            onChange={(e) => setSearchValue(e.target.value.toLowerCase())}
          />
        </Grid>
      </Grid>

      <Table size="small">
        <TableHead>
          <TableRow>
            <TableCell size="small" padding="none">
              <Button
                type="button"
                size="small"
                endIcon={
                  sortConfig.key === "last_name" ? (
                    sortConfig.direction === "ascending" ? (
                      <ArrowDownwardIcon />
                    ) : (
                      <ArrowUpwardIcon />
                    )
                  ) : (
                    ""
                  )
                }
                onClick={() => requestSort("last_name")}
              >
                Last name
              </Button>
            </TableCell>
            <TableCell size="small" padding="none">
              <Button
                type="button"
                size="small"
                endIcon={
                  sortConfig.key === "first_name" ? (
                    sortConfig.direction === "ascending" ? (
                      <ArrowDownwardIcon />
                    ) : (
                      <ArrowUpwardIcon />
                    )
                  ) : (
                    ""
                  )
                }
                onClick={() => requestSort("first_name")}
              >
                First name
              </Button>
            </TableCell>
            <TableCell size="small" padding="none">
              Gender
            </TableCell>
            <TableCell size="small" padding="none">
              <Button
                size="small"
                type="button"
                endIcon={
                  sortConfig.key === "city" ? (
                    sortConfig.direction === "ascending" ? (
                      <ArrowDownwardIcon />
                    ) : (
                      <ArrowUpwardIcon />
                    )
                  ) : (
                    ""
                  )
                }
                onClick={() => requestSort("city")}
              >
                City
              </Button>
            </TableCell>
            <TableCell size="small" padding="none">
              <Button
                size="small"
                type="button"
                endIcon={
                  sortConfig.key === "country" ? (
                    sortConfig.direction === "ascending" ? (
                      <ArrowDownwardIcon />
                    ) : (
                      <ArrowUpwardIcon />
                    )
                  ) : (
                    ""
                  )
                }
                onClick={() => requestSort("country")}
              >
                Country
              </Button>
            </TableCell>
            <TableCell align="right" size="small" padding="none">
              Score
            </TableCell>
          </TableRow>
        </TableHead>
        <TableBody>
          {sortedData
            .filter(
              (item) =>
                searchValue === "" ||
                item.last_name?.toLowerCase().includes(searchValue) ||
                item.first_name?.toLowerCase().includes(searchValue) ||
                item.city?.toLowerCase().includes(searchValue) ||
                item.country?.toLowerCase().includes(searchValue)
            )
            .slice(0, readmore ? -1 : 10)
            .map((row) => (
              <TableRow key={row.id}>
                <TableCell>{row.last_name}</TableCell>
                <TableCell>{row.first_name}</TableCell>
                <TableCell>{row.gender}</TableCell>
                <TableCell>{row.city}</TableCell>
                <TableCell>{row.country}</TableCell>
                <TableCell align="right">{row.score}</TableCell>
              </TableRow>
            ))}
        </TableBody>
      </Table>
      <div className={classes.seeMore}>
        <Link onClick={() => setReadmore(!readmore)} color="primary" href="#">
          {readmore ? "See less scores" : "see more scores"}
        </Link>
      </div>
    </React.Fragment>
  );
};

export default ScoreTable;

ScoreTable.propTypes = {
  userData: PropTypes.array.isRequired,
};
