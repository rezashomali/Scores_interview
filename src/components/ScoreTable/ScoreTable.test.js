import React from "react";
import renderer from "react-test-renderer";
import ScoreTable from "./ScoreTable";

describe("ScoreTable", () => {
  test("renders properly", () => {
    const data = [
      {
        city: "Olofstr√∂m",
        country: "SE",
        created_at: "2017-04-05T02:28:37Z",
        email: "mledstone0@mayoclinic.com",
        first_name: "Murdock",
        gender: "Male",
        id: "a67e6828-99d9-4d8d-9cd7-8aff12e95973",
        last_name: "Ledstone",
        score: 100,
      },
    ];
    const tree = renderer.create(<ScoreTable userData={data} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
