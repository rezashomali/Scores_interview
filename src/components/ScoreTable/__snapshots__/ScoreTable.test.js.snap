// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScoreTable renders properly 1`] = `
Array [
  <div
    className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3"
  >
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6"
    >
      <h2
        className="MuiTypography-root MuiTypography-h6 MuiTypography-colorPrimary MuiTypography-gutterBottom"
      >
        Scores listing
      </h2>
    </div>
    <div
      align="right"
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6"
    >
      <div
        className="MuiFormControl-root MuiTextField-root"
      >
        <label
          className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-marginDense MuiInputLabel-outlined"
          data-shrink={false}
        >
          search...
        </label>
        <div
          className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl MuiInputBase-marginDense MuiOutlinedInput-marginDense"
          onClick={[Function]}
        >
          <input
            aria-invalid={false}
            autoFocus={false}
            className="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMarginDense MuiOutlinedInput-inputMarginDense"
            disabled={false}
            onAnimationStart={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required={false}
            type="text"
            value=""
          />
          <fieldset
            aria-hidden={true}
            className="PrivateNotchedOutline-root-2 MuiOutlinedInput-notchedOutline"
          >
            <legend
              className="PrivateNotchedOutline-legendLabelled-4"
            >
              <span>
                search...
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
    </div>
  </div>,
  <table
    className="MuiTable-root"
    role={null}
  >
    <thead
      className="MuiTableHead-root"
      role={null}
    >
      <tr
        className="MuiTableRow-root MuiTableRow-head"
        role={null}
      >
        <th
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-head MuiTableCell-paddingNone MuiTableCell-sizeSmall"
          scope="col"
        >
          <button
            className="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textSizeSmall MuiButton-sizeSmall"
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={0}
            type="button"
          >
            <span
              className="MuiButton-label"
            >
              Last name
              
            </span>
          </button>
        </th>
        <th
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-head MuiTableCell-paddingNone MuiTableCell-sizeSmall"
          scope="col"
        >
          <button
            className="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textSizeSmall MuiButton-sizeSmall"
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={0}
            type="button"
          >
            <span
              className="MuiButton-label"
            >
              First name
              
            </span>
          </button>
        </th>
        <th
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-head MuiTableCell-paddingNone MuiTableCell-sizeSmall"
          scope="col"
        >
          Gender
        </th>
        <th
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-head MuiTableCell-paddingNone MuiTableCell-sizeSmall"
          scope="col"
        >
          <button
            className="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textSizeSmall MuiButton-sizeSmall"
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={0}
            type="button"
          >
            <span
              className="MuiButton-label"
            >
              City
              
            </span>
          </button>
        </th>
        <th
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-head MuiTableCell-paddingNone MuiTableCell-sizeSmall"
          scope="col"
        >
          <button
            className="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textSizeSmall MuiButton-sizeSmall"
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={0}
            type="button"
          >
            <span
              className="MuiButton-label"
            >
              Country
              
            </span>
          </button>
        </th>
        <th
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight MuiTableCell-paddingNone MuiTableCell-sizeSmall"
          scope="col"
        >
          Score
        </th>
      </tr>
    </thead>
    <tbody
      className="MuiTableBody-root"
      role={null}
    >
      <tr
        className="MuiTableRow-root"
        role={null}
      >
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall"
        >
          Ledstone
        </td>
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall"
        >
          Murdock
        </td>
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall"
        >
          Male
        </td>
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall"
        >
          Olofstr√∂m
        </td>
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall"
        >
          SE
        </td>
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeSmall"
        >
          100
        </td>
      </tr>
    </tbody>
  </table>,
  <div
    className="makeStyles-seeMore-1"
  >
    <a
      className="MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-colorPrimary"
      href="#"
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
    >
      see more scores
    </a>
  </div>,
]
`;
